File: lab5/java/comp26120/enum_solution.java
package comp26120;

import java.util.ArrayList;

public class enum_kp extends KnapSack {
    public enum_kp(String filename) {
	super(filename);
    }

    
    public static void main(String[] args) {
	enum_kp knapsack = new enum_kp(args[0]);

	knapsack.print_instance();
	knapsack.enumerate();
    }

    public void enumerate() {
	// Do an exhaustive search (aka enumeration) of all possible ways to pack
	// the knapsack.
	// This is achieved by creating every binary solution vector of length Nitems.
	// For each solution vector, its value and weight is calculated.

	int i; // item index
	ArrayList<Boolean> solution = new ArrayList<Boolean>(Nitems + 1); // Solution vector representing items packed
	solution.add(null); // C implementation has null first object
	    
	ArrayList<Boolean> best_solution = new ArrayList<Boolean>(Nitems + 1); // Solution vector for best solution found
	best_solution.add(null);
	
	int best_value; // total value packed in the best solution
	double j = 0;
	int infeasible; // 0 means feasible; -1 means infeasible (violates the capacity constraint)

	// set the knapsack initially empty
	for (i = 1; i <= Nitems; i++) {
	    solution.add(false);
	    best_solution.add(false);
	}

	QUIET=true;
	best_value = 0;

	while (!(next_binary(solution, Nitems))) {
	    j = j+1.0;
	    if (!QUIET)
		System.out.format("done %g of the full enumeration\n", j/Math.pow(2.0,Nitems));
	    infeasible=check_evaluate_and_print_sol(solution);  // calculates the value and weight and feasibility
	    if ((infeasible == 0) && total_value > best_value) {
		best_value = total_value;

		for (i = 1; i <= Nitems; i++) {
		    best_solution.set(i, solution.get(i));
		}
	    }
	    if (!QUIET) {
		System.out.format("best so far has value=%d\n", best_value);
	    }
	}
	QUIET = false;
	System.out.format("Finished.\nPack the following items for optimal\n");
	check_evaluate_and_print_sol(best_solution);
    }

    public boolean next_binary(ArrayList<Boolean> str, int Nitems) {
	// Called with an ArrayList of 0/1 entires with length Nitems
	// (0th item is null for some reason inherited from C implementation - I expect
	// because in C the array is here treated as a pointer to a string).
	// If we treat this array list as a binary number, then this
	// Function adds "1" - e.g., {0,0,0,1} would turn to {0,0,1,0}
	// It returns true when there are no possible binary numbers left 
	int i = Nitems;
	while(i>=1) {
	    if (str.get(i) == true) {
		str.set(i, false);
		i--;
	    } else {
		str.set(i, true);
		break;
	    }
	}
	if (i == 0) {
	    return true;
	} else {
	    return false;
	}
    }
}

File: lab5/c/enum_solution.c

// enum - full enumeration of knapsack solutions
// (C) Joshua Knowles, 2010-2013
// for issues: email j.knowles@manchester.ac.uk

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>

FILE *fp;  // file pointer for reading the input files
int Capacity;     // capacity of the knapsack (total weight that can be stored)
int Nitems;    // number of items available
int *item_weights;  // vector of item weights
int *item_values;  // vector of item profits or values
int *temp_indexes;  // list of temporary item indexes for sorting items by value/weight
int QUIET=0; // this can be set to 1 to suppress output

extern void read_knapsack_instance(char *filename);
extern void print_instance();
extern void sort_by_ratio();
extern int check_evaluate_and_print_sol(int *sol,  int *total_value, int *total_weight);
void enumerate();
int next_binary(int *str, int Nitems);


int main(int argc, char *argv[])
{
  read_knapsack_instance(argv[1]);
  print_instance();
  enumerate();
  return(0);
}

void enumerate()
{
  // Do an exhaustive search (aka enumeration) of all possible ways to pack
  // the knapsack.
  // This is achieved by creating every binary solution vector of length Nitems.
  // For each solution vector, its value and weight is calculated.


  int i;  // item index
  int solution[Nitems+1];   // (binary) solution vector representing items packed
  int best_solution[Nitems+1];  // (binary) solution vector for best solution found
  int best_value; // total value packed in the best solution
  double j=0;
  int total_value, total_weight; // total value and total weight of current knapsack solution
  int infeasible;  // 0 means feasible; -1 means infeasible (violates the capacity constraint)

  // set the knapsack initially empty
  for(i=1;i<=Nitems;i++)
    solution[i]=0;

  QUIET=1;

  best_value=0;
  while(!(next_binary(&solution[1], Nitems)))
    {
      j=j+1.0;
      if(!(QUIET))
	printf("done %g of the full enumeration\n", j/pow(2.0,Nitems));
      infeasible=check_evaluate_and_print_sol(solution, &total_value, &total_weight);  // calculates the value and weight and feasibility
      if((infeasible==0)&&(total_value>best_value))
	{
	  best_value=total_value;
	
	  for(i=1;i<=Nitems;i++)
	    best_solution[i]=solution[i];	
	}
      if(!(QUIET))
	printf("best so far has value=%d\n", best_value);
      
    }
  QUIET=0;
  printf("Finished.\nPack the following items for optimal\n");
  check_evaluate_and_print_sol(best_solution, &total_value, &total_weight);
  
}


int next_binary(int *str, int Nitems)
{
  // Called with a binary string of length Nitems, this 
  // function adds "1" to the string, e.g. 0001 would turn to 0010.
  // If the string overflows, then the function returns 1, else it returns 0.
  int i=Nitems;
  while(i>=0)
    {
      if(str[i]==1)
	{
	  str[i]=0;
	  i--;
	}
      else
	{
	  str[i]=1;
	  break;
	}
    }
  if(i==-1)
    return(1);
  else
    return(0);

}
